test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/HistoryTests.cs

```diff
@@ -120,58 +127,88 @@ public async Task GivenAValueForSince_WhenGettingSystemHistory_TheServerShouldRe
         {
             var tag = Guid.NewGuid().ToString();
 
-            var since = await GetStartTimeForHistoryTest(tag);
+            var since = await CreatePatientAndGetStartTimeForHistoryTest(tag);
             var sinceUriString = HttpUtility.UrlEncode(since.ToString("o"));
 
             Thread.Sleep(500);  // put a small gap between since and the first edits
-            _createdResource.Resource.Text = new Narrative { Div = "<div>Changed by E2E test</div>" };
+            _createdResource.Resource.Text = new Narrative { Div = $"<div>Changed by E2E test {tag}</div>" };
 
             var updatedResource = await CreateResourceWithTag(_createdResource.Resource, tag);
 
-            using FhirResponse<Bundle> readResponse = await _client.SearchAsync($"_history?_tag={tag}&_since={sinceUriString}&_count={count}");
+            FhirResponse<Bundle> response;
+            List<Bundle.EntryComponent> readResponse = new List<Bundle.EntryComponent>();
+            string searchString = $"_history?_since={sinceUriString}";
+            string selfLink = searchString;
+            while (!string.IsNullOrEmpty(searchString))
+            {
+                response = await _client.SearchAsync(searchString);
+                var readResponseWithMatchingTag = response.Resource.Entry.Where(e => e.Resource.Meta.Tag.Any(t => t.Code == tag)).ToList();
 
-            var readResponseWithMatchingTag = readResponse.Resource.Entry.Where(e => e.Resource.Meta.Tag.Where(t => t.Code == tag).Any()).ToList();
-            AssertCount(1, readResponseWithMatchingTag);
+                readResponse.AddRange(readResponseWithMatchingTag);
+                searchString = response.Resource.NextLink?.ToString();
 
-            var obsHistory = readResponseWithMatchingTag[0].Resource as Observation;
+                if (readResponseWithMatchingTag.Count == 1)
```

Comment:
Just checking my understanding, are we adding this if-statement to make sure we find the self link that points to the page that has the resource with the matching tag?

Reply:
Sorry, last follow up question! Do we need to iterate over subsequent pages when doing the self link search on line 163 if the link already points the page with the matching tag?

Reply:
No problem! I think we do need to iterate over the pages. SelfLink looks like this - http://localhost/_history?_since=2022-02-24T05%3A19%3A51.7570000%2B00%3A00. Also, the tag doesn't really work so there isn't a guarantee to get the right result in one iteration. For example, check the SS below 
![image](https://user-images.githubusercontent.com/86023331/155462951-82014f48-d04f-4a28-a6bf-ea74f8395255.png)
Before we query for the selfLink I added a bunch of resources and when we query for the first time the matching tag result is not returned. (check the watch for matching tag expression)

Reply:
As discussed, the self link doesn't point to a specific page. It will be set to the initial query string, and we are simply setting it when we find the resource with the matching tag.

Reply:
Backtracking a bit here: each page's self link _does_ point back to that specific page. Every self link except for the self link of the first page of results will not consider new resources that are created after the initial request.

A test I carried out:

1. Created a number of patients
2. Submitted the request: `GET <fhir-server>/Patient/_history`, noting the self link on the first page is set to `GET <fhir-server>/Patient/_history`
3. Used the next link to navigate to the second page of results
4. Saved the self link on the second page of results, noting it includes a continuation token
5. Created more patients
6. Used the self link saved in step 4, noting it points back to that specific page
7. Used the next link to navigate to the third page of results, noting the patients created in step 5 do not show up